### PR TITLE
[WIP] IMG-245 add NiFi processor

### PR DESCRIPTION
**Pull request by glenhein**
The NiFi processor extracts NITF metadata and create flow file attributes with names based on NITF taxonomy.

### CODE REVIEW

File: extract/src/main/java/org/codice/imaging/nitf/extract/ImageSegmentHandler.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    try {
+      final TargetId targetId = imageSegment.getImageTargetId();
+      if (targetId != null) {
+        final String targetIdValue = targetId.textValue();
+        if (StringUtils.isNotBlank(targetIdValue)) {
+          return targetIdValue;
+        }
+      }
+    } catch (NitfFormatException e) {
+      LOGGER.debug(
```
Comment (bellcc) [inline extract/src/main/java/org/codice/imaging/nitf/extract/ImageSegmentHandler.java]: ✏️ Include the exception in the log statement

File: extract/target/maven-archiver/pom.properties [hunk 2]
```diff
@@ -0,0 +1,5 @@
+#Generated by Maven
```
Comment (bellcc) [inline extract/target/maven-archiver/pom.properties]: ✏️ I think we need to update the `.gitignore` to not include files from `target`

File: extract/pom.xml [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    </parent>
+
+    <artifactId>nitf-extract</artifactId>
+    <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-api</artifactId>
+            <version>1.8.0</version>
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/pom.xml]: OBE: the nifi related code is moving to standalone project

File: extract/pom.xml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/pom.xml]: None of the POMs in imaging-nitf have the codice header. Maybe @dcruver knows if there is a specific reason.

File: extract/src/main/java/org/codice/imaging/nitf/extract/BaseExtract.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import org.apache.commons.lang.StringUtils;
+import org.codice.imaging.nitf.core.tre.Tre;
+
+public abstract class BaseExtract implements Function<Tre, Map<String, String>> {
+
+  private final String treName;
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/BaseExtract.java]: The name of the TRE (Tagged Record Extension, it's a nitf thing).

File: extract/src/main/java/org/codice/imaging/nitf/extract/ExtractACFTB.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+package org.codice.imaging.nitf.extract;
+
+import java.util.Arrays;
+import java.util.List;
+
+class ExtractACFTB extends BaseExtract {
+
+  private static final List<String> ATTRIBUTES =
+      Arrays.asList(
+          "AC_MSN_ID",
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ExtractACFTB.java]: I'll have to think about this. Downstream projects, like the nifi nitf processor, would be interested in the constants, but only as the appear as attributes in the nifi flow, such as, "nitf.ACFTB.AC_MSN_ID".

File: extract/src/main/java/org/codice/imaging/nitf/extract/ExtractCSDIDA.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ *
+ */
+package org.codice.imaging.nitf.extract;
+
+import java.util.Arrays;
+import java.util.List;
+
+class ExtractCSDIDA extends BaseExtract {
+  private static final List<String> ATTRIBUTES =
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ExtractCSDIDA.java]: fixed

File: extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  private static final LabelSegmentHandler LABEL_SEGMENT_HANDLER =
+      new LabelSegmentHandler((segment, metadata) -> {});
+
+  public Map<String, String> extract(InputStream inputStream) throws IOException {
+    Map<String, String> metadata = new HashMap<>();
+    extract(inputStream, metadata);
+    return metadata;
+  }
+
+  public void extract(InputStream inputStream, Map<String, String> metadata) throws IOException {
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java]: That's something I'm currently thinking about. I don't really like the side-effect of populating a map. I'd rather convert everything to return a map and then have all of those maps merged.

File: extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+  private static void extractTreMetadata(
+      Map<String, String> metadata, TaggedRecordExtensionHandler taggedRecordExtensionHandler) {
+    taggedRecordExtensionHandler.getTREsRawStructure().getTREs().stream()
+        .map(ExtractMetadata::applyTreHandler)
+        .forEach(metadata::putAll);
+  }
+
+  private static Map<String, String> applyTreHandler(Tre tre) {
+    return TRE_HANDLERS.getOrDefault(tre.getName().trim(), IGNORE).apply(tre);
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java]: I'd rather not muck with the Tre ctor. There are downstream users of imaging-nitf and I wouldn't want to break something they are expecting. At this point, I was just trying to reproduce the logic from alliance. Maybe @dcruver has an opinion on whether or not it is safe to modify the Tre ctor.

File: extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    return metadata;
+  }
+
+  public void extract(InputStream inputStream, Map<String, String> metadata) throws IOException {
+    NitfSegmentsFlow flow;
+
+    try {
+      flow = new NitfParserInputFlowImpl().inputStream(inputStream).allData();
+    } catch (NitfFormatException e) {
+      throw new IOException("Unable to get the nitf segments flow.", e);
```
[the remark this answers is not in the document]
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ExtractMetadata.java]: I'm generally opposed to logging and throwing at the same time. I've always been taught that you do one or the other, but not both. But I'm open to discussion on the topic!

File: extract/src/main/java/org/codice/imaging/nitf/extract/ImageSegmentHandler.java (hunk 1 above)
Comment (glenhein) [reply extract/src/main/java/org/codice/imaging/nitf/extract/ImageSegmentHandler.java]: fixed

**Comment by glenhein** [comment]
@bradh The duplicate file will be going away. An XML configuration is interesting, and I'll explore it. I'll think about how to add some kind of indexing or naming to support multiple occurrences of the same segment or TRE.

**Comment by bradh** [comment]
> @bradh An XML configuration is interesting, and I'll explore it. I'll think about how to add some kind of indexing or naming to support multiple occurrences of the same segment or TRE.
We already have the XML configuration, and parsing logic. 

We also already have the code to handle multiple values inside a TRE (like a looping structure).

**Pull request closed without merging**